### PR TITLE
Add method to calculate holidays for duration of year

### PR DESCRIPTION
**Pull request by jonathanpike**
I needed a way of getting a list of all the holidays for a region from a certain date to the end of the year, so I wrote it.  I used some of Rails' Date calculations in order to do this (ie. end_of_month, change).  I thought that if I needed this functionality, it may also be useful for others.

### CODE REVIEW

**Comment by ptrimble** [comment]
Wow, thank you! I'll take a look at this over the coming week before I leave this weekend and let you know if I have any comments.

**Comment by ptrimble** [comment]
Sorry @jonathanpike, things got away from me this week and I don't know if I will be able to get to this before I leave for a weekend trip. I'll probably need to get to this at the beginning of next week.

Sorry again!

**Comment by jonathanpike** [comment]
@ptrimble No worries, take your time!  I'm just happy that I can contribute to a gem that made my life easier!  Let me know if you have any comments or suggestions when you do get the chance to review it.

File: lib/holidays/core_extensions/date.rb [hunk 1]
```diff
@@ -29,6 +29,24 @@ def holiday?(*options)
         holidays && !holidays.empty?
       end
 
+      # Returns a new Date where one or more of the elements have been changed according to the +options+ parameter.
+      # The +options+ parameter is a hash with a combination of these keys: <tt>:year</tt>, <tt>:month</tt>, <tt>:day</tt>.
+      #
+      #   Date.new(2007, 5, 12).change(day: 1)               # => Date.new(2007, 5, 1)
+      #   Date.new(2007, 5, 12).change(year: 2005, month: 1) # => Date.new(2005, 1, 12)
+      def change(options)
```
Comment (ptrimble) [inline lib/holidays/core_extensions/date.rb]: 😍 

Oh my goodness. The PR is worth it *just* for this set of additions. Very very nice!

File: test/test_holidays.rb [hunk 2]
```diff
@@ -139,6 +139,12 @@ def test_any_region_holiday_next
     end
   end
 
+  def test_any_region_holiday_year
+    holidays = Holidays.year_holidays([:ca_on], Date.civil(2016, 2, 23))
+    holidays.each {|h| puts h}
+    # assert_equal 12, holidays.length
```
Comment (ptrimble) [inline test/test_holidays.rb]: Oops! Can you uncomment this?

File: test/test_helper.rb [hunk 3]
```diff
@@ -11,12 +11,17 @@
...
 require 'holidays'
 require 'holidays/core_extensions/date'
+require 'holidays/core_extensions/time'
 
 # Loads core extension for use in various definition tests as necessary
 class Date
   include Holidays::CoreExtensions::Date
 end
 
+class Time
```
Comment (ptrimble) [inline test/test_helper.rb]: This is a good addition, can you also add some specific tests around the new `Date` and `Time` core extensions? Right now those tests aren't great so feel free to create a new test file just to test the new fields. I would then probably refactor to migrate the core ext tests to your new file (if we go in that direction).

File: test/test_holidays.rb (hunk 2 above)
Comment (ptrimble) [inline test/test_holidays.rb]: In addition, I know that I am asking for a stretch, but can you please add a few additional tests? 

I'm trying to expand the general test coverage and I think that we should start covering edge cases for each piece of functionality.

Potential edge cases:

* Jan 1 (i.e. get all holidays for that year)
* Dec 31 (i.e. get the last holiday, or none if NYE is not considered a holiday in a particular region)
* A few random months/days in different parts of the year
* What happens when someone specifies Feb 29 on a non leap year?
* A few different years (1990, 2016, 2020, 2050, just random years)
* Call without a specified year (i.e. default of `Date.today`)
* Anything else I missed?

Each of these should be their own test. I'm not a huge fan of putting multiple test cases into one test function.

Again, I recognize that this is asking for a bit of a stretch and that we don't always do this for other tests but I think it A) wouldn't be that much additional effort and B) is the 'right' way to do things going forward.

Let me know what you think!

File: lib/holidays.rb [hunk 4]
```diff
@@ -160,6 +160,50 @@ def next_holidays(holidays_count, options, from_date = Date.today)
...
+    #       {name=>"Easter Sunday",...},
+    #       {:name=>"Victoria Day",...},
+    #       {:name=>"Canada Day",...},
+    #       {:name=>"Civic Holiday",...},
+    #       {:name=>"Labour Day",...},
+    #       {:name=>"Thanksgiving",...},
+    #       {:name=>"Remembrance Day",...},
+    #       {:name=>"Christmas Day",...},
+    #       {:name=>"Boxing Day",...}]
+    def year_holidays(options, from_date = Date.today)
```
Comment (ptrimble) [inline lib/holidays.rb]: @ttwo32 Brought up to me that this looks very similar to `next_holidays`. I do agree with that but...I don't see a clean way to merge the two. I am not a fan of some kind of branching logic (i.e. a flag that takes us down different paths depending on what the flag says). I'm of the opinion that a little duplication is better than trying to force together two similar ideas.

Anyone have any opinions on this? I'm inclined to leave it as-is for the time being.

File: lib/holidays.rb [hunk 5]
```diff
@@ -160,6 +160,50 @@ def next_holidays(holidays_count, options, from_date = Date.today)
...
+      raise ArgumentError if options.empty?
+      raise ArgumentError unless options.is_a?(Array)
+
+      # remove the timezone
+      from_date = from_date.new_offset(0) + from_date.offset if from_date.respond_to?(:new_offset)
+
+      from_date = get_date(from_date)
+      to_date = from_date.to_datetime.change(:month => 12).end_of_month.to_date
```
Comment (ptrimble) [inline lib/holidays.rb]: I like this but does this seems to require that a user add the core extension for `Date`. That's a change since we don't require the core extension otherwise as far as I know.

This is what I got when I tried to run locally:

```sh
% be rake console
irb -rubygems -I lib -r holidays.rb
irb(main):001:0> Holidays.year_holidays([:us])
NoMethodError: undefined method `change' for #<DateTime:0x00561d9fd08528>
        from /home/vagrant/dev/holidays/lib/holidays.rb:196:in `year_holidays'
        from (irb):1
        from /home/vagrant/.rbenv/versions/2.2.2/bin/irb:11:in `<main>'
irb(main):002:0>
```

This is a problem. Is there a way that we can get the last day of the year without requiring the core extension? Perhaps:

```ruby
Date.new(y=Date.today.year,m=12,d=31).yday
```

Or something along those lines?

**Comment by ptrimble** [comment]
@jonathanpike A have a few questions/comments:

* The core extension requirement in `year_holidays`. We should find another way to get the last day of the target year.
* Adding a few additional tests. As I mentioned in the comment I know that this is a stretch ask but I think now is a good time to expand our coverage of test cases.

Overall great! Very happy to receive this, once we address this minor concerns we'll be good to merge and release. 👍

**Comment by ptrimble** [comment]
Oh and thanks for your patience. Obviously took me longer to get to this than I anticipated. 😄

**Comment by jonathanpike** [comment]
@ptrimble I've added additional tests for year holidays, removed the requirement for the core extension, and also added some tests (in a new file) for the Date and Time core extensions.  Let me know what you think, and if any further changes are needed.  I really appreciate your detailed review!

File: lib/holidays.rb (hunk 4 above)
Comment (ttwo32) [inline lib/holidays.rb]: @jonathanpike I agree with you and thank you for reading next_holidays. :D

Comment (ptrimble) [inline lib/holidays.rb]: Nope, I 100% agree that a little duplication is better than a bad combination. Leave it as-is. 👍

File: test/holidays/core_extensions/test_date_time.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+class Date
+  include Holidays::CoreExtensions::Date
+end
+
+class Time 
+  include Holidays::CoreExtensions::Time
+end 
+
+class CoreExtensionDateTimeTests < Test::Unit::TestCase
```
Comment (ptrimble) [inline test/holidays/core_extensions/test_date_time.rb]: 😍

**Comment by ptrimble** [comment]
Fantastic. I'll get this merged and pushed up shortly. Thanks so much @jonathanpike! We really appreciate it, great PR. 👍

**Comment by jonathanpike** [comment]
:rocket: This is officially my first open source contribution.  Thanks for maintaining such a useful gem -- you guys made my life so much easier!

**Comment by ttwo32** [comment]
@jonathanpike Congratulation for your first OSS contribution! I think that this PR is a useful feature!! Thank you!! 😄
